Cost functions now support Stan Math, Kept the previous classes for backward compatability.

FaroukY opened the pull request:
We can now write any arbitrary cost functions in Stan Math, and get the gradients with respect to all variables, regardless how complex the cost function is. This PR has the class FirstOrderSAGCostFunctionInterface which acts as an interface for building Stochastic Average Cost functions. I've also added a unit test to show how to use the class. All reference values where validated on Tensorflow.

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,178 @@
+/*
```
you should use the shorter version of the license, see for example SGObject.h

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+/*
```
should use the shorter version of the license

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+#include <vector>
+
+
+namespace shogun{
```
whitespace

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+#define FIRSTORDERSAGCOSTFUNCTIONINTERFACE_H
+
+#include <stan/math.hpp>
+#include <Eigen/Dense>
```
we never include eigen directly in shogun headers, only in linalg but even then it's always via shogun/mathematics/eigen3.h

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+   * where \f$(y_i,x_i)\f$ is the i-th sample,
+   * \f$y_i\f$ is the label and \f$x_i\f$ is the features
+   */
+  class FirstOrderSAGCostFunctionInterface : public FirstOrderSAGCostFunction{
```
i guess this class is never gonna be part of an SGObject, hence we could make a pass over all the anti-shogun-patterns that you've used below

vigsterkr replied:
SG_ADD is not used anywhere in this class...

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+  public:
+    /** Consturctor that initializes X, y, trainable_parameters, and the cost_function */
+  	FirstOrderSAGCostFunctionInterface(
+  		Eigen::Matrix<float64_t, Eigen::Dynamic, Eigen::Dynamic>* X,
```
why not SGVector<float64_t>?

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+    /** Consturctor that initializes X, y, trainable_parameters, and the cost_function */
+  	FirstOrderSAGCostFunctionInterface(
+  		Eigen::Matrix<float64_t, Eigen::Dynamic, Eigen::Dynamic>* X,
+  		Eigen::Matrix<float64_t, 1, Eigen::Dynamic>* y,
```
why not SGVector<float64_t>?

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+  	FirstOrderSAGCostFunctionInterface() {};
+
+  	/** Setter for the training data X */
+  	virtual void set_training_data(
```
these args coudl be SGVectors

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+    virtual void set_cost_function(std::function < stan::math::var(std::vector<stan::math::var>*) >* total_cost);
+
+
+  	~FirstOrderSAGCostFunctionInterface();
```
override/virtual?

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+  	 */
+  	virtual float64_t get_cost();
+
+    /** Get the sample size
```
indenting.... plz use clang-format

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+     *
+     * @return the sample size
+     */
+    virtual int32_t get_sample_size();
```
index_t

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+  	Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1>* m_trainable_parameters;
+
+    /** cost_for_ith_point is the cost contributed by each point in the training data */
+    std::vector< std::function < stan::math::var(int32_t) > >* m_cost_for_ith_point;
```
indent

vigsterkr commented on tests/unit/optimization/FirstOrderSAGCostFunctionInterface_unittest.cc:
```diff
@@ -0,0 +1,184 @@
+/*
+* Copyright (c) The Shogun Machine Learning Toolbox
```
use the shorter license

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+
+  protected:
+  	/** X is the training data in column major matrix format */
+  	Eigen::Matrix<float64_t, Eigen::Dynamic, Eigen::Dynamic>* m_X;
```
you have not added these to the param framework, nor this will be ever be able to be added... you need to use SGVector for that

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+  	Eigen::Matrix<float64_t, Eigen::Dynamic, Eigen::Dynamic>* m_X;
+
+  	/** y is the ground truth, or the correct prediction */
+  	Eigen::Matrix<float64_t, 1, Eigen::Dynamic>* m_y;
```
same as above... SGVector

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+  	Eigen::Matrix<float64_t, 1, Eigen::Dynamic>* m_y;
+
+  	/** trainable_parameters are the variables that are optimized for */
+  	Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1>* m_trainable_parameters;
```
this currently is foobar... you wont be able to serialize reproducible this class

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+    std::vector< std::function < stan::math::var(int32_t) > >* m_cost_for_ith_point;
+
+    /** total_cost is the total cost to be minimized, that in this case is a form of sum of cost_for_ith_point*/
+    std::function < stan::math::var(std::vector<stan::math::var>*) >* m_total_cost;
```
indent + problems with serialization

vigsterkr commented:
plz run clang-format on your code prior pushing into a PR

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunction.h:
```diff
@@ -96,23 +96,23 @@ class FirstOrderSAGCostFunction
 	virtual SGVector<float64_t> get_gradient()=0;
 
-	/** Get the cost given current target variables 
+	/** Get the cost given current target variables
```
could you adjust your editor so it doesnt cause all this noise?

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,178 @@
+   Eigen::Matrix<float64_t, 1, Eigen::Dynamic>* y_new
+ )
+ {
+  REQUIRE(X_new!=NULL, "X must be non NULL");
```
Is this a user facing error?
If no, pls say "No X provided.\n" (also missing newline)

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,178 @@
+   std::vector< std::function < stan::math::var(int32_t) > >* new_cost_f
+ )
+ {
+   REQUIRE(new_cost_f, "The cost function must be a non NULL vector of stan variables");
```
see above, and also for all other error msgs

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,178 @@
+SGVector<float64_t> FirstOrderSAGCostFunctionInterface::get_gradient()
+{
+  int32_t num_of_variables = m_trainable_parameters->rows();
+  REQUIRE(num_of_variables > 0, "Number of training parameters must be greater than 0");
```
Pls always print the provided number
"Number of training parameters (%d) must be positive.\n"
https://github.com/shogun-toolbox/shogun/wiki/Assertions

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,178 @@
+  int32_t n = get_sample_size();
+  REQUIRE(n>0 , "Number of sample must be greater than 0");
+
+  for(auto i=0; i<n; ++i)
```
minor: `for (auto i : range(n))` for all range based loops

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,178 @@
+        auto grad = get_gradient();
+        for(auto j=0; j<params_num; ++j)
+        {
+          average_gradients[j]+= (grad[j]/n);
```
can you use linalg here pls, rather than those vanilla loops

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,173 @@
+
+  	/** Get the SAMPLE gradient value wrt target variables
+  	 *
+  	 * WARNING
```
I dont understand this warning thing. Could you make it more clear?

karlnapf commented on tests/unit/optimization/FirstOrderSAGCostFunctionInterface_unittest.cc:
```diff
@@ -0,0 +1,184 @@
+}
+
+
+TEST(LeastSquareTestCostFunction, ONALINE)
```
whats ONALINE?
Pls try to do slightly more elaborate test names :)

FaroukY replied:
Sorry, meant that the points are on a line (ONALINE)

karlnapf replied:
points_on_a_line

karlnapf commented on tests/unit/optimization/FirstOrderSAGCostFunctionInterface_unittest.cc:
```diff
@@ -0,0 +1,167 @@
+	X(0, 2) = 1;
+	X(1, 0) = 0;
+	X(1, 1) = 1;
+	X(1, 2) = 2;
```
is there a chance of re-using this test data or the general setup? If so it would be good to put it into a function or a fixture. It also makes the test smaller

iglesias replied:
++

FaroukY commented:
Alright, addressed all changes.

iglesias commented on tests/unit/optimization/FirstOrderSAGCostFunctionInterface_unittest.cc:
```diff
@@ -0,0 +1,146 @@
+using stan::math::var;
+using std::function;
+
+/** This is a temporary fix. The variables are now variables
```
What does this mean?

vigsterkr replied:
we need to figure out how we can store (if we want a none primitive typed std::vector, like stan::math:var)... hence the temporary story... just add a FIXME/TODO term there and then its good

FaroukY replied:
@iglesias  As Viktor said. The original cost function needs a *reference* to an SGVector<float64_t> of the parameters. The parameters have now been changed in the cost function to SGVector<stan::math::var> so we're going to have to change the interface of the minimizers before returning a SGVector<stan::math::var>,

iglesias commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,161 @@
+using Eigen::Dynamic;
+using Eigen::Matrix;
+using stan::math::var;
+using std::function;
```
It does not feel good to have type aliases in headers. They also become effective in places where the header is included, no?

vigsterkr replied:
second that this actually contaminates any file that will include this header

FaroukY replied:
I thought that if you use using std::function; in a header, then you'd need to access it from another file which has the include as FirstOrderSAGCostFunctionInterface::function no? In other words, shouldn't it be under the scope of the .h file?

FaroukY replied:
Actually you're right, I just forgot my c++! Including it in the header file does include it in other files as well. I'll get rid of it.

iglesias commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,161 @@
+{
+	/** @brief The first order stochastic cost function base class for
+	 * implementing
+	 *  The SAG Cost function
```
Style: newline (after implementing) and capitalization (implementing \n The) are odd.

FaroukY replied:
Sorry, I ran clang-format on it and didn't check the styling. Will fix the styling of the comments

iglesias commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,161 @@
+
+		/** Default constructor, use setter helpers to set X, y,
+		*trainable_parameters, and
+		*		cost_function
```
This is not informative doc :-) I'd remove very general comments like this one, they add noise.

iglesias commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,161 @@
+
+	protected:
+		/** X is the training data in column major matrix format */
+		SGMatrix<float64_t>* m_X;
```
Why are all these members pointers?

FaroukY replied:
@iglesias Just a convention (I try to pass things by reference/pointer instead of value to avoid copying). But I agree I will get rid of SGMatrix's pointers since it already implements that under the hood.

iglesias commented on src/shogun/optimization/FirstOrderSAGCostFunction.h:
```diff
@@ -31,8 +31,8 @@
 
```
As you are already touching this file, it would be nice to do a tiny extra effort and update the Copyright to the more modern shorter version.

iglesias commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,161 @@
+	 * where \f$(y_i,x_i)\f$ is the i-th sample,
+	 * \f$y_i\f$ is the label and \f$x_i\f$ is the features
+	 */
+	class FirstOrderSAGCostFunctionInterface : public FirstOrderSAGCostFunction
```
It seems a bit odd that the interface inherits from the class.

FaroukY replied:
Hmm, how about FirstOrderSAGArbitraryCostFunction ??

iglesias commented:
I have the feeling that @vigsterkr's first general comment is not addressed. @FaroukY, did you discuss with anybody about it?

iglesias commented:
How when it is needed, I think. We can sync in a few days about what
attributes these cost functions should have, which ones make sense to make
part of the public api.

For instance right now I see that the Interface class even contains the
training data.

On 25 May 2018 at 09:33, Viktor Gal <notifications@github.com> wrote:

> *@vigsterkr* commented on this pull request.
> ------------------------------
>
> In tests/unit/optimization/FirstOrderSAGCostFunctionInterface_unittest.cc
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r190811707>
> :
>
> > + *
> + * Authors: Elfarouk
> + */
> +
> +#include <gtest/gtest.h>
> +#include <shogun/optimization/FirstOrderSAGCostFunctionInterface.h>
> +#include "FirstOrderSAGCostFunctionInterface_unittest.h"
> +#include <Eigen/Dense>
> +
> +using namespace shogun;
> +using Eigen::Matrix;
> +using Eigen::Dynamic;
> +using stan::math::var;
> +using std::function;
> +
> +/** This is a temporary fix. The variables are now variables
>
> we need to figure out how we can store (if we want a none primitive typed
> std::vector, like stan::math:var)... hence the temporary story... just add
> a FIXME/TODO term there and then its good
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r190811707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABGrdqb2zbTIE2Z2t7vyIb87A8-3wD2lks5t17O6gaJpZM4UIEvl>
> .
>

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+
+bool FirstOrderSAGCostFunctionInterface::next_sample()
+{
+	auto num_of_samples = get_sample_size();
```
there's just no need for this variable on the stack... and just extra line of code. just do:

```
if (m_index_of_sample >= get_sample_size())
	return false; 
```

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+
+SGVector<float64_t> FirstOrderSAGCostFunctionInterface::get_gradient()
+{
+	int32_t num_of_variables = m_trainable_parameters->rows();
```
auto

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+
+float64_t FirstOrderSAGCostFunctionInterface::get_cost()
+{
+	int32_t n = get_sample_size();
```
auto

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+
+SGVector<float64_t> FirstOrderSAGCostFunctionInterface::get_average_gradient()
+{
+	int32_t params_num = m_trainable_parameters->rows();
```
auto

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+	int32_t params_num = m_trainable_parameters->rows();
+	SGVector<float64_t> average_gradients(params_num);
+
+	int32_t old_index_sample = m_index_of_sample;
```
auto

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+	SGVector<float64_t> average_gradients(params_num);
+
+	int32_t old_index_sample = m_index_of_sample;
+	int32_t n = get_sample_size();
```
auto

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+	for (index_t i = 0; i < n; ++i)
+	{
+		m_index_of_sample = i;
+		auto grad = get_gradient();
```
do we need this variable?

vigsterkr replied:
just do `average_gradients += get_gradient()`

vigsterkr commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+
+FirstOrderSAGCostFunctionInterface::FirstOrderSAGCostFunctionInterface(
+    SGMatrix<float64_t>* X, SGMatrix<float64_t>* y,
+    Matrix<var, Dynamic, 1>* trainable_parameters,
```
what's the story behind these being pointers instead of ref?

FaroukY replied:
I am not sure what is the convention in Shogun. But I always used pointers in my previous C++ coding (Just to avoid the syntax of reference). If you'd like, I think we can change it to reference here.

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+	int32_t num_of_variables = m_trainable_parameters->rows();
+	REQUIRE(
+	    num_of_variables > 0,
+	    "Number of training parameters must be greater than 0");
```
missing newline. Also pls always print the probided number
"Number of training parameters (%d) must be positive.\n"

FaroukY replied:
Sure, but in this case, I'll just print: "Number of training parameters must be greater than 0, you provided 0".... since you can't have num_of_variables<0

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+	f_i.grad();
+
+	for (auto i = 0; i < num_of_variables; ++i)
+		gradients[i] = (*m_trainable_parameters)(i, 0).adj();
```
can't this be done without a loop using `std::transform` or something?

FaroukY replied:
Didn't know this, but Eigen has unaryExpr which allows you to apply any function to all elements in the matrix. So I've removed the vanilla loop and replaced it with a call to that function.

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp:
```diff
@@ -0,0 +1,149 @@
+{
+	int32_t n = get_sample_size();
+	Matrix<var, Dynamic, 1> cost_argument(n);
+	for (auto i = 0; i < n; ++i)
```
as abovek would be better to do this without loop (if easy)
If not, at least do a `for (auto i : range(n))`

FaroukY replied:
This one is a bid harder to do without a loop. Since we're calling a function inside and passing an argument, Eigen::unaryExpr wouldn't work.

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,161 @@
+		 * cost_function */
+		FirstOrderSAGCostFunctionInterface(
+		    SGMatrix<float64_t>* X, SGMatrix<float64_t>* y,
+		    Matrix<var, Dynamic, 1>* trainable_parameters,
```
could we maybe alias things like `Matrix<var, Dynamic, 1>` into something more readable?

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,161 @@
+		 * */
+		virtual bool next_sample();
+
+		/** Get the SAMPLE gradient value wrt target variables
```
why is this SAMPLE? rather than sample

FaroukY replied:
Just wanted to emphasise it. Changed it to lower case.

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,161 @@
+
+		/** Get the SAMPLE gradient value wrt target variables
+		 *
+		 * WARNING
```
not sure what the warning is supposed to mean, I commented this before but maybe I missed your reply?

FaroukY replied:
@karlnapf It means we should return the partial derivative of the ith sample with respect to the variables, and not the partial derivative of the total cost with respect with respect to the variables. 

So for example, if f_i(w) = (w^t*X[i]-y[i])^2 is the error of the ith data point, then this function returns the partial derivative of the **current** data sample with respect to w, rather than returning the derivative of Sum_i(f_i(w)) with respect to w.

This function (and get_average_gradient()) is from the base class (FirstOrderSAGCostFunction) and had the warning in its docs. Do you want me to change the warning in both files to something else, and if so, what do you want to clarify?

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,161 @@
+		 */
+		virtual index_t get_sample_size();
+
+		/** Get the AVERAGE gradient value wrt target variables
```
average

FaroukY commented on tests/unit/optimization/FirstOrderSAGCostFunctionInterface_unittest.h:
```diff
@@ -0,0 +1,35 @@
+ */
+
+#include <shogun/lib/config.h>
+#ifndef FIRSTORDERSAGCOSTFUNCTIONINTERFACE_UNITTEST_H
```
Not sure to be honest. I looked at the other files and they all had <one word with no spaces>_UNITTEST_H so I followed the convention. If there is no guideline I'll add more _ between words

FaroukY commented on src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h:
```diff
@@ -0,0 +1,161 @@
+
+		/** total_cost is the total cost to be minimized, that in this case is a
+		 * form of sum of cost_for_ith_point*/
+		function<var(Matrix<var, Dynamic, 1>*)>* m_total_cost;
```
@iglesias  Not silly at all! I actually thought about this, but I wanted to separate the logic of calculating the cost function from the class so that base class can support any cost function. We can also make a pure virtual method and let the user implement it. Both work.

FaroukY commented:
Alright, 2nd iteration. Comments are either 1) Addressed or 2) I replied to the comment for some clarification

iglesias commented:
UNITTEST is good. It is about the FIRSTORDERSAG...

On Fri, May 25, 2018, 23:17 Elfarouk Yasser <notifications@github.com>
wrote:

> *@FaroukY* commented on this pull request.
> ------------------------------
>
> In tests/unit/optimization/FirstOrderSAGCostFunctionInterface_unittest.h
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191011586>
> :
>
> > @@ -0,0 +1,35 @@
> +/*
> + * This software is distributed under BSD 3-clause license (see LICENSE file).
> + *
> + * Authors: Elfarouk
> + */
> +
> +#include <shogun/lib/config.h>
> +#ifndef FIRSTORDERSAGCOSTFUNCTIONINTERFACE_UNITTEST_H
>
> Not sure to be honest. I looked at the other files and they all had
> _UNITTEST_H so I followed the convention. If there is no guideline I'll add
> more _ between words
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191011586>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABGrdr1DEkwVl5kykUr3oKnV_b1tBTtpks5t2HT8gaJpZM4UIEvl>
> .
>

iglesias commented:
No, the concept is much simpler :-) If you want to use std::function
anywhere (header or implementation file), you just need to include the
functional from the stl.

On Fri, May 25, 2018, 23:19 Elfarouk Yasser <notifications@github.com>
wrote:

> *@FaroukY* commented on this pull request.
> ------------------------------
>
> In src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191011757>
> :
>
> > +#ifndef FIRSTORDERSAGCOSTFUNCTIONINTERFACE_H
> +#define FIRSTORDERSAGCOSTFUNCTIONINTERFACE_H
> +
> +#include <stan/math.hpp>
> +#include <shogun/mathematics/eigen3.h>
> +#include <shogun/lib/config.h>
> +#include <shogun/optimization/FirstOrderSAGCostFunction.h>
> +#include <functional>
> +#include <vector>
> +#include <shogun/lib/SGMatrix.h>
> +#include <shogun/lib/SGVector.h>
> +
> +using Eigen::Dynamic;
> +using Eigen::Matrix;
> +using stan::math::var;
> +using std::function;
>
> I thought that if you use using std::function; in a header, then you'd
> need to access it from another file which has the include as
> FirstOrderSAGCostFunctionInterface::function no? In other words, shouldn't
> it be under the scope of the .h file?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191011757>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABGrdi0xavZtBk734HlIcclGP2jDPdQ8ks5t2HVEgaJpZM4UIEvl>
> .
>

iglesias commented:
Making it a member attribute does not separate from the class, right?

On Fri, May 25, 2018, 23:37 Elfarouk Yasser <notifications@github.com>
wrote:

> *@FaroukY* commented on this pull request.
> ------------------------------
>
> In src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191014764>
> :
>
> > +		/** X is the training data in column major matrix format */
> +		SGMatrix<float64_t>* m_X;
> +
> +		/** y is the ground truth, or the correct prediction */
> +		SGMatrix<float64_t>* m_y;
> +
> +		/** trainable_parameters are the variables that are optimized for */
> +		Matrix<var, Dynamic, 1>* m_trainable_parameters;
> +
> +		/** cost_for_ith_point is the cost contributed by each point in the
> +		 * training data */
> +		Matrix<function<var(int32_t)>, Dynamic, 1>* m_cost_for_ith_point;
> +
> +		/** total_cost is the total cost to be minimized, that in this case is a
> +		 * form of sum of cost_for_ith_point*/
> +		function<var(Matrix<var, Dynamic, 1>*)>* m_total_cost;
>
> @iglesias <https://github.com/iglesias> Not silly at all! I actually
> thought about this, but I wanted to separate the logic of calculating the
> cost function from the class so that base class can support any cost
> function. We can also make a pure virtual method and let the user implement
> it. Both work.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191014764>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABGrdsejxOrMTsv2UnsowdjF5uPwrXcYks5t2HmIgaJpZM4UIEvl>
> .
>

iglesias commented:
That makes sense of course, avoiding unnecessaey copies. But here nothing
is passed, this is just the definition or declaration of the attributes of
a class.

Having these as pointers may make sense. It is of course not always the
wrong choice. Although one should prefer using smart instead of naked
pointers.

Conceptually, whether an attribute should be a pointer, reference, or value
is a question of ownership and life times.

On Fri, May 25, 2018, 23:38 Elfarouk Yasser <notifications@github.com>
wrote:

> *@FaroukY* commented on this pull request.
> ------------------------------
>
> In src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191015041>
> :
>
> > +		 * WARNING
> +		 * This method returns
> +		 * \f$ \frac{\sum_i^n{ \frac{\partial f_i(w) }{\partial w} }}{n}\f$
> +		 *
> +		 * For least squares, that is the value of
> +		 * \f$ \frac{\frac{\partial f(w) }{\partial w}}{n} \f$ given \f$w\f$ is
> +		 * known
> +		 * where \f$f(w)=\frac{ \sum_i^n{ (y_i-w^t x_i)^2 } }{2}\f$
> +		 *
> +		 * @return average gradient of target variables
> +		 */
> +		virtual SGVector<float64_t> get_average_gradient();
> +
> +	protected:
> +		/** X is the training data in column major matrix format */
> +		SGMatrix<float64_t>* m_X;
>
> @iglesias <https://github.com/iglesias> Just a convention (I try to pass
> things by reference/pointer instead of value to avoid copying). But I agree
> I will get rid of SGMatrix's pointers since it already implements that
> under the hood.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191015041>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABGrdu-0BhPwSEU-J71nDFwBP-J0IJyxks5t2HnygaJpZM4UIEvl>
> .
>

iglesias commented:
It sounds like you may have been writing mostly C and using a C++ compiler
:-)

Fairly common practice unfortunately. There are more expressive and just
plain better ways of using C++.

On Sat, May 26, 2018, 00:50 Elfarouk Yasser <notifications@github.com>
wrote:

> *@FaroukY* commented on this pull request.
> ------------------------------
>
> In src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191025127>
> :
>
> > +/*
> + * This software is distributed under BSD 3-clause license (see LICENSE file).
> + *
> + * Authors: Elfarouk
> + */
> +
> +#include <shogun/optimization/FirstOrderSAGCostFunctionInterface.h>
> +#include <shogun/mathematics/Math.h>
> +using namespace shogun;
> +using stan::math::var;
> +using std::function;
> +using Eigen::Matrix;
> +
> +FirstOrderSAGCostFunctionInterface::FirstOrderSAGCostFunctionInterface(
> +    SGMatrix<float64_t>* X, SGMatrix<float64_t>* y,
> +    Matrix<var, Dynamic, 1>* trainable_parameters,
>
> I am not sure what is the convention in Shogun. But I always used pointers
> in my previous C++ coding (Just to avoid the syntax of reference). If you'd
> like, I think we can change it to reference here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191025127>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABGrdpx8eIDup6vB1-E0XcGLRGnDJBHgks5t2Iq6gaJpZM4UIEvl>
> .
>

iglesias commented:
Probably not under zero, unless there's a nasty bug corrupting the memory.
I see tht way of being over-careful about writing the error message a form
of defensive programming.

On Sat, May 26, 2018, 00:51 Elfarouk Yasser <notifications@github.com>
wrote:

> *@FaroukY* commented on this pull request.
> ------------------------------
>
> In src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191025301>
> :
>
> > +
> +bool FirstOrderSAGCostFunctionInterface::next_sample()
> +{
> +	auto num_of_samples = get_sample_size();
> +	if (m_index_of_sample >= num_of_samples)
> +		return false;
> +	++m_index_of_sample;
> +	return true;
> +}
> +
> +SGVector<float64_t> FirstOrderSAGCostFunctionInterface::get_gradient()
> +{
> +	int32_t num_of_variables = m_trainable_parameters->rows();
> +	REQUIRE(
> +	    num_of_variables > 0,
> +	    "Number of training parameters must be greater than 0");
>
> Sure, but in this case, I'll just print: "Number of training parameters
> must be greater than 0, you provided 0".... since you can't have
> num_of_variables<0
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/shogun-toolbox/shogun/pull/4294#discussion_r191025301>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABGrdoCL2i48HnB6FGTUE2boWNpmeUf8ks5t2IsCgaJpZM4UIEvl>
> .
>

karlnapf commented:
What is the state here?

FaroukY commented:
@karlnapf This will be a [WIP] pr. The aim is to provide a very minimal working optimization mini module that works only on stan (Will support 1 minimizer, 1 updater, 1 type of cost functions, and so on. Then other minimizers, updaters, .... will be updated sequentially one by one once we have a "Framework" we agree on. I'll be pushing a new update here very very soon.

vigsterkr commented on src/shogun/optimization/StanFirstOrderSAGCostFunction.cpp:
```diff
@@ -0,0 +1,140 @@
+	    m_trainable_parameters->unaryExpr(
+	        [](stan::math::var x) -> float64_t { return x.adj(); });
+	// clone needed because gradients is local variable
+	return SGVector<float64_t>(gradients).clone();
```
you could just simply wrap the EigenVectorXt with SGVector and not do cloning...
namely
```
SGVector<float64>(gradients.data(), gradients.size())
```

but i reckon it'd be better that simply just create an
```
SGVector<float64_t> gradients(size);
SGVector<float64_t>::EigenVectorXt mapped_gradients = gradients;
/// now you can use mapped_gradients as an eigen vector

return gradients;
```

vigsterkr commented on src/shogun/optimization/StanFirstOrderSAGCostFunction.cpp:
```diff
@@ -0,0 +1,140 @@
+
+StanFirstOrderSAGCostFunction::StanFirstOrderSAGCostFunction(
+    SGMatrix<float64_t> X, SGMatrix<float64_t> y,
+    StanVector* trainable_parameters,
```
still dont get why we need pointers.

vigsterkr commented on src/shogun/optimization/StanFirstOrderSAGCostFunction.cpp:
```diff
@@ -0,0 +1,140 @@
+	    m_trainable_parameters->unaryExpr(
+	        [](stan::math::var x) -> float64_t { return x.adj(); });
+	// clone needed because gradients is local variable
+	return SGVector<float64_t>(gradients).clone();
```
you could just simply wrap the EigenVectorXt with SGVector and not do cloning...
namely
```
SGVector<float64>(gradients.data(), gradients.size())
```
but i reckon it'd be better that simply just create an

```
SGVector<float64_t> gradients(size);
SGVector<float64_t>::EigenVectorXt mapped_gradients = gradients;
/// now you can use mapped_gradients as an eigen vector

return gradients;
```

FaroukY replied:
@vigsterkr  But the data was being destroyed since the variable was local. I tried the second one but when I returned gradients, it just contained garbage values?

vigsterkr replied:
if you return SGVector<float64_t> that should trigger a copy-ctor that should ++ the ref counter and hence shouldn't delete the data itself.

vigsterkr replied:
there should be many examples where this is used in shogun.... e.g. https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/machine/gp/GaussianLikelihood.cpp#L238

vigsterkr commented on src/shogun/optimization/StanFirstOrderSAGCostFunction.cpp:
```diff
@@ -0,0 +1,140 @@
+		m_index_of_sample = i;
+		average_gradients += get_gradient();
+	}
+	average_gradients.scale(1.0 / n);
```
plz use `linalg::scale()`

vigsterkr commented on src/shogun/optimization/StanFirstOrderSAGCostFunction.cpp:
```diff
@@ -0,0 +1,140 @@
+
+SGVector<float64_t> StanFirstOrderSAGCostFunction::get_average_gradient()
+{
+	int32_t params_num = m_trainable_parameters->rows();
```
auto

vigsterkr commented on src/shogun/optimization/StanFirstOrderSAGCostFunction.cpp:
```diff
@@ -0,0 +1,140 @@
+
+	for (index_t i = 0; i < n; ++i)
+	{
+		m_index_of_sample = i;
```
mmmm this feels very very strange... as this definitely makes the whole cost function not thread-safe. i'm not saying that they should be, but i'm not convinced atm that this is actually required, or the only way to solve what you want

FaroukY replied:
This is one of the things I'm discussing with you today.

FaroukY replied:
This should be addressed in the next set of commits.

vigsterkr commented on tests/unit/optimization/StanFirstOrderSAGCostFunction_unittest.cc:
```diff
@@ -0,0 +1,139 @@
+	auto f_i = [X, y](StanVector* w, int32_t idx) {
+		var wx_y = (*w)(0, 0) * X(0, idx) + (*w)(1, 0) * X(1, idx) - y(0, idx);
+		var res = wx_y * wx_y;
+		res /= 2;
```
nitpicking, but this is the same as `res *= 0.5` except that the former is slower

vigsterkr commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,175 @@
+		var wx_y = (*w)(0, 0) * X(0, idx) + (*w)(1, 0) * X(1, idx) +
+		           (*w)(2, 0) * X(2, idx) - y(0, idx);
+		var res = wx_y * wx_y;
+		res /= 2;
```
same as above

vigsterkr commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,175 @@
+
+	auto total_cost = get_tot_cost();
+
+	SquareErrorTestCostFunction* fun = new SquareErrorTestCostFunction(
```
these all could be auto right?

vigsterkr commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,175 @@
+	SquareErrorTestCostFunction* fun = new SquareErrorTestCostFunction(
+	    X, y, &w, &cost_for_ith_point, &total_cost);
+
+	SGDMinimizer* opt = new SGDMinimizer(fun);
```
auto

vigsterkr commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,175 @@
+
+	SGDMinimizer* opt = new SGDMinimizer(fun);
+
+	ConstLearningRate* rate = new ConstLearningRate();
```
auto

vigsterkr commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,175 @@
+	ConstLearningRate* rate = new ConstLearningRate();
+	rate->set_const_learning_rate(0.01);
+
+	GradientDescendUpdater* updater = new GradientDescendUpdater();
```
auto

iglesias commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,175 @@
+
+void RegressionData::init()
+{
+	// The data is referenced from Wu Lin Unit tests
```
Would it make sense to have it in a common place?

FaroukY replied:
@iglesias Do you mean to put his name in a more obvious place? If so then I've already put his name in the header (copyrights). No plagiarism here :)

iglesias replied:
No. That comment refers to all the lines below, where x and y are initialized.

FaroukY replied:
@iglesias Do you mean you'd like the data to be in a more clear place (Like the top) rather than in the middle?

iglesias commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,175 @@
+	opt->set_gradient_updater(updater);
+	opt->set_learning_rate(rate);
+
+	int32_t num_passes = 20;
```
It is good to be consistent.

Above you are doing `rate->set_const_learning_rate(0.01)` and here instead you declare and define `num_passes` before passing it to the setter.

iglesias commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,175 @@
+	return f_i;
+}
+
+function<var(StanVector*)> get_tot_cost()
```
Can this `StanVector*` be a `const StanVector&` instead?

iglesias commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,175 @@
+function<var(StanVector*)> get_tot_cost()
+{
+	auto cost = [](StanVector* v) {
+		var total_cost = v->sum();
```
If kept as *, there must be a check for nullptr.

iglesias commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,175 @@
+	        Matrix<function<var(StanVector*, float64_t)>, Dynamic, 1>::Constant(
+	            n, 1, f_i);
+
+	auto total_cost = get_tot_cost();
```
Nitpicking, it'd make sense to just have the lambda here.

FaroukY replied:
Agreed, its only used once here so no need to define a function that wraps the lambda.

iglesias commented on tests/unit/optimization/StanFirstOrderSAGCostFunction_unittest.cc:
```diff
@@ -0,0 +1,139 @@
+}
+
+function<var(StanVector*, float64_t)>
+cost_for_ith_datapoint(SGMatrix<float64_t>& X, SGMatrix<float64_t>& y)
```
Ok. So, I would like to understand this pattern based on declaring a function that returns a std::function, and the function's definition is a lambda. It feels like it is over-complicated.

Why not simply defining a function? Please explain.

FaroukY replied:
Just refactoring. I was using the same definition of a lambda twice in the file, so instead I just wrapped the lambda definition in a function: (cost_for_ith_datapoint) and used it in both places that were defining the lambdas so there is less copy paste.

iglesias replied:
Just use a function that returns a 'var' (instead of std::function). For the definition of the function just use the logic in the body of the lambda.

FaroukY commented on src/shogun/optimization/StanFirstOrderSAGCostFunction.cpp:
```diff
@@ -0,0 +1,113 @@
+		average_gradients += get_gradient();
+	}
+  average_gradients.scale(1.0/n);
+  //lingalg::scale(average_gradients, 1.0/n);
```
@vigsterkr  Do I need to set the gpu backend before using linalg::scale ?

iglesias commented on src/shogun/optimization/StanFirstOrderSAGCostFunction.h:
```diff
@@ -0,0 +1,177 @@
+	 * where \f$(y_i,x_i)\f$ is the i-th sample,
+	 * \f$y_i\f$ is the label and \f$x_i\f$ is the features
+	 */
+	class StanFirstOrderSAGCostFunction : public FirstOrderSAGCostFunction
```
Not sure if this relationship between `StanFirstOrderSAGCostFunction` and `FirstOrderSAGCostFunction` makes sense.

FaroukY replied:
Agreed. I've changed the parent class to be FirstOrderStochasticCostFunction, since all these implemented functions are directly from it. (FirstOrderSAGCostFunction also inherits from FirstOrderStochasticCostFunction so it makes sense here to inherit from it as the Stan version is an alternative of FirstOrderSAGCostFunction).

iglesias commented on src/shogun/optimization/StanFirstOrderSAGCostFunction.h:
```diff
@@ -0,0 +1,177 @@
+		SGMatrix<float64_t> m_X;
+
+		/** y is the ground truth, or the correct prediction */
+		SGMatrix<float64_t> m_y;
```
`FirstOrderSAGCostFunction` does not have members for the training data. Why are they needed here?

FaroukY replied:
@iglesias In the unit test of FirstOrderSAGCostFunction, we implement an additional class called CRegressionExample that wraps FirstOrderSAGCostFunction and contains the training data. I've simply removed the necessity for that class since it just acted as a wrapper around the cost function and data, and have included the data in the loss function, or atleast a reference to the data.

iglesias replied:
We must be careful and we must no add unnecessary relationships. If it turns out that the training data is a member both in the cost function, and in the neural network, and in etc etc, that is going to cause lot of usage confusion.

It does not sounds unreasonable that the tests have a facility packaged to prepare input data and avoid code duplication (e.g. inside a class a the CRegressionExample you are mentioning). What did you find wrong with it?

FaroukY commented on src/shogun/optimization/StanFirstOrderSAGCostFunction.h:
```diff
@@ -0,0 +1,177 @@
+		SGMatrix<float64_t> m_y;
+
+		/** trainable_parameters are the variables that are optimized for */
+		StanVector& m_trainable_parameters;
```
I was just thinking of it as a design problem. As in aggregation vs composition. The cost function doesn't own the parameters, so I felt aggregation works better here (implemented in C++ as a reference) What do you think?

iglesias replied:
I think a reference member should not be used but maybe it is ok.

iglesias replied:
Any news about this concern?

iglesias commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,167 @@
+	var x1(0), x2(0), x3(0);
+	w(0, 0) = x1;
+	w(1, 0) = x2;
+	w(2, 0) = x3;
```
The syntax to initialize this `StanVector` is a bit awkward. Can you do better? For example with something std::vector-like it could just look like `std::vector<var>{x1, x2, x3}`.

FaroukY replied:
Hmm, well StanVector is just an alias for Eigen<var, Dynamic, 1> so I am not sure I can change the constructor unless I define a wrapper class for Eigen<var, Dynamic, 1>.

iglesias replied:
Please have a look at initializing Eigen::Matrix in Eigen documentation.

https://eigen.tuxfamily.org/dox/group__TutorialAdvancedInitialization.html

iglesias commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,167 @@
+}
+
+function<var(const StanVector&, float64_t)>
+cost_for_ith_datapoints(SGMatrix<float64_t>& X, SGMatrix<float64_t>& y)
```
Refactor.

FaroukY replied:
Will be refactoring the stan tests once the neural network is working

iglesias commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,167 @@
+
+	auto f_i = cost_for_ith_datapoints(X, y);
+
+	Matrix<function<var(const StanVector&, float64_t)>, Dynamic, 1>
```
This is a case where auto really makes sense actually.

FaroukY replied:
I wanted to cast here, since the output is not exactly:

Matrix<function<var(const StanVector&, float64_t)>, Dynamic, 1>

iglesias replied:
What do you mean?

iglesias commented on tests/unit/optimization/StanStochasticMinimizers_unittest.cc:
```diff
@@ -0,0 +1,167 @@
+		return total_cost;
+	};
+
+	auto fun = new SquareErrorTestCostFunction(
```
No deletion for this one and the CostLearningRate?

FaroukY replied:
@iglesias They're handled in the minimizer's destructor.

iglesias commented on src/shogun/neuralnets/StanNeuralLogisticLayer.h:
```diff
@@ -0,0 +1,53 @@
+ * [squared error measure](http://en.wikipedia.org/wiki/Mean_squared_error) is
+ * used
+ */
+class StanNeuralLogisticLayer : public StanNeuralLinearLayer
```
Perhaps it is only a naming issue, but saying that a LogisticLayer is a LinearLayer sounds odd.

iglesias replied:
Any plans to fix?

iglesias commented on src/shogun/neuralnets/StanNeuralLogisticLayer.h:
```diff
@@ -0,0 +1,53 @@
+	/** default constructor */
+	StanNeuralLogisticLayer();
+
+	/** Constuctor
```
Minor typo: constructor.

iglesias commented on src/shogun/neuralnets/StanNeuralLogisticLayer.h:
```diff
@@ -0,0 +1,53 @@
+			CDynamicObjectArray* layers);
+
+
+	virtual const char* get_name() const { return "NeuralLogisticLayer"; }
```
Missing Stan prefixed in the name?

iglesias commented on src/shogun/neuralnets/StanNeuralNetwork.h:
```diff
@@ -0,0 +1,195 @@
+	 */
+	virtual void initialize_neural_network(float64_t sigma = 0.01f);
+
+	virtual ~CNeuralNetwork();
```
Did you mean destructor of `StanNeuralNetwork`?

iglesias commented on src/shogun/neuralnets/StanNeuralNetwork.h:
```diff
@@ -0,0 +1,195 @@
+	/** apply machine to data in means of regression problem */
+	virtual CRegressionLabels* apply_regression(CFeatures* data);
+	/** apply machine to data in means of multiclass classification problem */
+	virtual CMulticlassLabels* apply_multiclass(CFeatures* data);
```
I am not sure these ones should be here. Let's check with @karlnapf or @vigsterkr.

iglesias commented on src/shogun/neuralnets/StanNeuralNetwork.cpp:
```diff
@@ -0,0 +1,280 @@
+ }
+
+
+ SGVector<float64_t>* StanNeuralNetwork::get_layer_parameters(int32_t i)
```
This looks strange. The signature in the header is `StanVector& get_layer_parameters(int32_t i);`.

FaroukY commented:
@iglesias 

I previously discussed with Viktor that I'd first finish writing the whole Neural Network module (Which is not already done), then write the unit tests for it and make sure the logic is correct, and then we can go back and refactor all the points addressed previously. I've already addressed some points mentioned, but the main focus was getting a MVP of the Neural network running only on stan for gradient calculation. I'm not ignoring the points, but I'm just more concerned now with finishing the actual logic (Then I can refactor/change names and do the other required changes). 

I'm currently writing the unit tests for the whole neural network code, and this will be the last file to be added to this PR. It should be a standalone PR changing to a tested NN to use stan only.

iglesias commented:
Hi @FaroukY! That's all right. Though I find that a ~2000 lines patch for a so-called MVP has a bit of over-kill smell.

Still, I do suggest you to keep in mind the comments that have already been made. Just think that you don't want to build upon something that will need to be completely changed from its foundations. I am not necessarily saying that is the case here, but I believe you should keep it mind.
